Imported document about undefined behavior and safe api in LLD (#127)
rustfoundation/safety-critical-rust-consortium: subcommittee/coding-guidelines/initiatives/safe-use-of-unsafe-guidelines/rust-embedded-lld-safe-definition.md

considered regarding undefined behavior.

Microcontroller low level driver should provide safe API only
considering just the microcontroller but excluding the voltage regulator
because they should be generic and usable for any system.

<p align="center">
<img src="./rust-embedded-lld-safe-definitions-images/safe_lld_fig3.png"  alt="Figure 3 Scope of microcontroller LLD" height="200">
</p>

Voltage regulator driver requires a SPI low level driver in order and
their combination shall protect from undefined behavior that could arise
from microcontroller + voltage regulator. Voltage regulator driver is
supposed to be generic and independent of the device connected to the
output of voltage regulator.
<p align="center">
<img src="./rust-embedded-lld-safe-definitions-images/safe_lld_fig4.png"  alt="Figure 4 Scope of voltage regulator driver and microcontroller LLD" height="200">
</p>
Finally, system mode manager API shall guarantee the freedom from
undefined behavior of full system as described in Figure 1 HW system.

## Conclusions

In this document I tried to show why the measures required to have safe
LLD are not clear. The reason derives from the definition of “undefined

> behavior”. This is a call for Rust community and Rust specification

Reviewer: From my point of view, we should be careful to delineate those things which the Rust Project and most of the Rust community would say lead to "undefined behavior" from a particular (and important!) segment which cares about safety-critical and writing low-level drivers.

I think that this involves some advocacy on our part as the Safety-Critical Rust Consortium Coding Guidelines Subcommittee.

As we have chatted about a few times, this seems to be a bit of a "scoping" or "system" issue where as you show here it depends on how you view the scope of your system and how it can do things which lead to "undefined behavior".

I'd recommend rewriting this section here to emphasize perhaps this "community" nature, since as you highlight the svd2rust crate here (which is maintained by the Tools team) is a first major collaborator in this mission you're embarking on.

Author: @PLeVasseur This was an old document that I though could be useful as food for thought and I have no problem to change.

The main point that I tried to highlight is that there is an inconsistency in the community regarding the definition of undefined behaviors and when use or not use `unsafe` keyword.

This inconsistency in the scope of safety this inconsistency can lead to different interpretation and  therefore safety risk.

I would propose to rewrite the document in following way:

* Highlight the inconsistency/unclarity of `unsafe` usage and undefined behavior definition across Rust community 
* Remove any reference to svd2rust
* Remove all reference to HW peripheral (I will mention only interface to C Low Level Driver) and consider a more generic external world (FFI)
* Remove the section about solution proposal and call for a community solution or better definition in Rust upcoming specification.
* Highlight the risk of this ambiguity in the context of safety application.

Is this addressing your recommendation ?

Reviewer: I think you've got the start of a useful document here to drive discussion, for sure.

The aspect that I think you address pretty well and could be due more emphasis is what we often look at in a safety-critical context, which is of system scoping, i.e. where the safety issues could crop up.

While in a vacuum calling a certain API that's FFI or directly accessing memory or manipulating hardware may not cause a safety concern from the point of view, it's within the greater context of what a system is trying to accomplish that something may result in undefined behavior. Your diagrams and explanations I think go to that point.

> The main point that I tried to highlight is that there is an inconsistency in the community regarding the definition of undefined behaviors and when use or not use unsafe keyword.

My main challenge with the document as-is is in this framing. I think that within the context of the Rust programming language the understanding of undefined behavior is ~reasonably well-defined.

However, when we take a function call that's doing something which in and of itself doesn't cause undefined behavior, but due to the ordering of the function called with respect to another function, that's where this system scoping framing can be useful.

e.g.

* Calling a `foo()` function: defined
* Calling `foo()`, then `bar()` functions: defined
* Calling `bar()`, then `foo()` functions: undefined behavior

Therefore, when doing a low-level driver that should accomplish `foo()` + `bar()` we will consider the system scope as surrounding both of them to ensure they're called in the correct order.

I think it's admirable to say that low-level software that's still fairly close to manipulating hardware should not be `unsafe`! I'd like to make sure it's being motivated and communicated in this way.

writers to have an unambiguous and well-defined description of
“undefined behavior” that includes HW drivers.

> I described a proposal inspired by ISO23232 out of context element that

Reviewer: ```suggestion
I described a proposal inspired by ISO26262 out of context element that
```

> # Rust Embedded Low Level Drivers and Safe Definition

Reviewer: Would be good to put images under `rust-embedded-lld-safe-definitions-images` so that if we have other documents with images they can go under a corresponding folder.

Author: Done

*An unsafe operation is an operation that may result in undefined
behavior that is not diagnosed as a static error. Unsafe operations are
referred to as unsafe Rust.*

This definition is aligned with what is written in Rustnomicon[^2]:

*No matter what, Safe Rust can't cause Undefined Behavior.*

Moreover, the keyword `unsafe` is used in multiple context[^3]:

  - *unsafe fn*: calling this function means abiding by a contract the
    compiler cannot enforce.

  - *unsafe trait*: implementing the trait means abiding by a contract
    the compiler cannot enforce.

  - *unsafe {}*: the contract necessary to call the operations inside
    the block has been checked by the programmer and is guaranteed to be
    respected.

  - *unsafe impl*: the contract necessary to implement the trait has
    been checked by the programmer and is guaranteed to be respected.

In this discussion only unsafe blocks and unsafe functions. Unsafe
functions can be called only in unsafe blocks and the user shall ensure
to use unsafe functions fulfilling all requirements described in the
documentation to avoid undefined behavior.

This simplifies code review and bug identification. The developers shall

> inspect first unsafe blocks in case application shows an undefined

Reviewer 2: ```suggestion
inspect first unsafe blocks in case the application shows an undefined
```

behavior. However, this doesn`t exclude that a function developer forgot
to mark a function as “unsafe”.

### What is “undefined behavior” in Rust? 

Ferrocene specification defines it as “unbounded error”[^4] or in other
words: everything could happen.

Rust Reference manual provide a non-exhaustive list of “undefined

> behaviors”[^5] The following behavior are worth to mentioned because it

Reviewer 2: ```suggestion
behaviors”[^5] The following behavior are worth mentioning because it
```

involves hardware:

  - *Executing code compiled with platform features that the current
    platform does not support (see target\_feature), except if the
    platform explicitly documents this to be safe.*

> Before to discuss safe/unsafe in Low Level Drivers let`s see the meaning

Reviewer 2: ```suggestion
Before discussing safe/unsafe in Low Level Drivers let`s see the meaning
```

of “undefined behavior” in the context of microcontroller/microprocessor
hardware.

## Hardware undefined behavior

> In many data sheet or user manual of micro-controller and microprocessor
> is used the term *undefined behavior* and in these documents is
> described how to avoid the conditions that could generate an undefined
> behavior.

Reviewer 2: I wasn't quite sure what this paragraph get getting at. Does this suggestion make sense, or are you trying to make a different point?

```suggestion
In the data sheets or user manuals of many micro-controllers and 
microprocessors the term *undefined behavior* is used mostly to describe
how to avoid the conditions that could generate an undefined behavior.
```

Author: It is the same. Thank you

What does mean *undefined behavior* in the context of microcontrollers
or microprocessor?

  - undefined behavior is not tested in simulation and in production.
    

>       - Every behavior could happen.

Reviewer 2: ```suggestion
      - Any behavior could happen.
```

process manufacturing, operating temperature…). Therefore, undefined
behavior is the worst-case scenario and there is no software test that
can detect it.

It shall be highlighted that an undefined behavior could be triggered
not only by wrong value of registers but also by wrong order of reading
and writing of registers.

## Rust safe/unsafe are not well defined when hardware is involved.

From previous discussion it could appear obvious that safe low level
drivers shall protect from any misuses that could trigger undefined
behavior even from hardware.

Unfortunately, this is not generally accepted and there are different
understandings. Let`s list them:

### Svd2rust

Svd2Rust[^6] generated peripheral access crates (PAC) from xml register
description. Peripheral access crates provides API to write and read
peripheral registers.

Safe functions provided by generated crates protect users from following
issue:

  - Writing out of range value for bitfields.

  - Concurrent access to same register from different interrupt/thread

  - Read or write registers that cannot be read or written.

_These functions cannot protect from undefined

> behavior triggered by improper order of calling functions_.

Reviewer 2: This is an interesting note, and something I hadn't considered.

    a very low level access to external HW function is considered safer
    than accessing a C library.

On svd2rust and safety there is an ongoing discussion in an issue of
GitHub project[^8] related to this topic.

### Is safe and unsafe not applicable to external world?

I have talked with some developers, and they don`t consider relevant HW
undefined behavior because are not part of Rust developed software. In
other words, they assume that unsafety shall be restricted to only rust
code memory issues, thread issue but not to HW component.

This position is similar to the one regarding the safety of *fs::open*
in Unix system[^9]. The *fs::open* is considered safe function even if
it can open special file */proc/self/mem* that allows unchecked full
access to process memory. Moreover, it is stated the following in the
documentation:

*Rust`s safety guarantees only cover what the program itself can do, and
not what entities outside the program can do to it.*

However, I still see the same conflicting usage of unsafe as in the
previous chapter:

  - C API are by default unsafe. Are not referring to external C
    program? If the previous definition holds, they should be considered

>     safe.

Reviewer 2: My thought here would be that because as you defined for the unsafe block,

> the contract necessary to call the operations inside the block has been checked by the programmer and is guaranteed to be respected

Any calls made via C API should be considered default unsafe until a programmer checks to ensure the contract of the API is fulfilled and Rust's safety guarantees will be met.

Author: Sorry but I don't fully understand this comment.
I am trying to explain that there is a contradiction between the statement:

> *Rust`s safety guarantees only cover what the program itself can do, and
not what entities outside the program can do to it.*

and the fact that FFI (external word) *are* considered unsafe by the compiler. 
Moreover the definition of unsafe block is copied from [here](https://doc.rust-lang.org/std/keyword.unsafe.html#the-different-meanings-of-unsafe)

Are you proposing a change in the text ?

Reviewer 2: I was not proposing a change in the text. I just didn't understand that you were setting up a contradiction there.

I understand what you are saying now, after reading your reply. Perhaps you could consider making that section more explicit, or you could just leave it as is.

Author: I improved the description

  - HW undefined behavior is already included in Rust Reference:
    “*Executing code compiled with platform features that the current
    platform does not support*”.

  - Measures required for safe DMA API in Embednomicon. DMA is external
    to Rust program but measures are taken to have safe API.
    
## Proposal

I think it is quite clear that a clear unambiguous and widely accepted
definition of undefined behavior in Rust language especially when HW is
involved is missing.

> While waiting a widely better definition in coming Rust specification, I
> would propose the following to solve this issue:

Reviewer 2: ```suggestion
While waiting for a vastly better definition in an upcoming Rust
specification, I would propose the following to solve this issue:
```

</p>

For each software component shall be documented which part of system is
considered regarding undefined behavior.

Microcontroller low level driver should provide safe API only
considering just the microcontroller but excluding the voltage regulator
because they should be generic and usable for any system.

<p align="center">
<img src="./rust-embedded-lld-safe-definitions-images/safe_lld_fig3.png"  alt="Figure 3 Scope of microcontroller LLD" height="200">
</p>

Voltage regulator driver requires a SPI low level driver in order and
their combination shall protect from undefined behavior that could arise
from microcontroller + voltage regulator. Voltage regulator driver is
supposed to be generic and independent of the device connected to the
output of voltage regulator.
<p align="center">
<img src="./rust-embedded-lld-safe-definitions-images/safe_lld_fig4.png"  alt="Figure 4 Scope of voltage regulator driver and microcontroller LLD" height="200">
</p>
Finally, system mode manager API shall guarantee the freedom from
undefined behavior of full system as described in Figure 1 HW system.

## Conclusions

In this document I tried to show why the measures required to have safe

> LLD are not clear. The reason roots down to the definition of “undefined

Reviewer 2: ```suggestion
LLD are not clear. The reason derives from the definition of “undefined
```

behavior”. This is a call for Rust community and Rust specification
writers to have an unambiguous and well-defined description of
“undefined behavior” that includes HW drivers.

I described a proposal inspired by ISO23232 out of context element that
is applicable to LLD and still preserve one of the main advantages of
Rust:

*No matter what, Safe Rust can't cause Undefined Behavior.*

[^1]: https://spec.ferrocene.dev/unsafety.html
[^2]: https://doc.rust-lang.org/nomicon/safe-unsafe-meaning.html

> [^3]: https://doc.rust-lang.org/std/keyword.unsafe.html\#the-different-meanings-of-unsafe

Author: Fixed

> [^4]: https://spec.ferrocene.dev/glossary.html\#undefined-behavior

Author: Fixed

[^5]: https://doc.rust-lang.org/reference/behavior-considered-undefined.html
[^6]: https://docs.rs/svd2rust/latest/svd2rust/
[^7]: https://docs.rust-embedded.org/embedonomicon/dma.html
[^8]: https://github.com/rust-embedded/svd2rust/issues/714

> [^9]: https://doc.rust-lang.org/std/os/unix/io/index.html\#procselfmem-and-similar-os-features

Author: Fixed

words: everything could happen.

Rust Reference manual provide a non-exhaustive list of “undefined
behaviors”[^5] The following behavior are worth mentioning because it
involves hardware:

  - *Executing code compiled with platform features that the current
    platform does not support (see target\_feature), except if the
    platform explicitly documents this to be safe.*

> Before discussing safe/unsafe in Low Level Drivers let`s see the meaning

Reviewer: ```suggestion
Before discussing safe/unsafe in Low Level Drivers let's see the meaning
```

Reviewer: I'll highlight this one here as I think my ask wasn't clear. It seems you're using tilde (`) instead of apostrophe (') in spots in the document. I was trying to note those spots. Could you switch the tildes for apostrophes where needed?

> *Rust`s safety guarantees only cover what the program itself can do, and

Reviewer: ```suggestion
*Rust's safety guarantees only cover what the program itself can do, and
```
